src/shapirowilk.jl

```diff
@@ -0,0 +1,250 @@
+    if firstindex(SWc.A) ≤ i ≤ lastindex(SWc.A)
+        return @inbounds SWc.A[i]
+    elseif isodd(SWc.N) && i == lastindex(SWc.A) + 1
+        return zero(eltype(SWc))
+    else
+        return -SWc.A[SWc.N+1-i]
+    end
+end
+
+function SWCoeffs(N::Int)
+    if N < 3
+        throw(ArgumentError("N must be greater than or equal to 3: $N"))
+    elseif N == 3 # exact
+        return SWCoeffs(N, [sqrt(2.0) / 2.0])
+    else
+        # Weisberg&Bingham 1975 statistic; store only positive half of m:
+        # it is (anti-)symmetric; hence '2' factor below
+        m = [-quantile(Normal(), (i - 3 / 8) / (N + 1 / 4)) for i in 1:div(N, 2)]
+        mᵀm = 2sum(abs2, m)
+        x = 1 / sqrt(N)
+        a₁ = m[1] / sqrt(mᵀm) + __RS92_C1(x) # aₙ = cₙ + (...)
+        if N ≤ 5
+            ϕ = (mᵀm - 2m[1]^2) / (1 - 2a₁^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1] = a₁
+        else
+            a₂ = m[2] / sqrt(mᵀm) + __RS92_C2(x) # aₙ₋₁ = cₙ₋₁ + (...)
+            ϕ = (mᵀm - 2m[1]^2 - 2m[2]^2) / (1 - 2a₁^2 - 2a₂^2)
+            m .= m ./ sqrt(ϕ) # A, but reusing m to save allocs
+            m[1], m[2] = a₁, a₂
+        end
+
+        return SWCoeffs(N, m)
+    end
+end
+
+function swstat(X::AbstractArray{<:Real}, A::SWCoeffs)
+    if last(X) - first(X) < length(X) * eps()
+        throw("data seems to be constant!")
+    end
+    AX = dot(A, X)
+    m = mean(X)
+    S² = sum(x -> abs2(x - m), X)
+    return AX^2 / S²
+end
+
+function pvalue(W::Real, A::SWCoeffs, N1=A.N)
+    A.N != N1 && throw("not implemented yet!")
+    if A.N == 3 # exact by Shapiro&Wilk 1965
+        return π / 6 * (asin(sqrt(W)) - asin(sqrt(0.75)))
+    elseif A.N ≤ 11
+        γ = __RS92_G(A.N)
+        if γ ≤ log(1 - W)
+            return zero(W)
+        end
+        w = -log(γ - log(1 - W))
+        μ = __RS92_C3(A.N)
+        σ = exp(__RS92_C4(A.N))
+    else
+        w = log(1 - W)
+        μ = __RS92_C5(log(A.N))
+        σ = exp(__RS92_C6(log(A.N)))
+    end
+    return ccdf(Normal(μ, σ), w)
+end
+
+struct ShapiroWilkTest <: HypothesisTest
+    SWc::SWCoeffs         # Expectation of order statistics for Shapiro-Wilk test
+    W::Float64            # test statistic
+    N1::Int               #
+end
+
+testname(::ShapiroWilkTest) = "Shapiro-Wilk normality test"
+population_param_of_interest(t::ShapiroWilkTest) =
+    ("Squared correlation of data and SWCoeffs (W)", 1.0, t.W)
+default_tail(::ShapiroWilkTest) = :left
+
+function show_params(io::IO, t::ShapiroWilkTest, indent)
+    l = 24
+    println(io, indent, rpad("number of observations:", l), t.SWc.N)
+    println(io, indent, rpad("censored ratio:", l), (t.SWc.N - t.N1) / t.SWc.N)
+    println(io, indent, rpad("W-statistic:", l), t.W)
+end
+
+pvalue(t::ShapiroWilkTest) = pvalue(t.W, t.SWc, t.N1)
+
+"""
+    ShapiroWilkTest(
+        X::AbstractArray{<:Real};
+        SWc::SWCoeffs=SWCoeffs(length(X)),
+        N1=length(X),
+        is_sorted=issorted(view(X, 1:N1))
+    )
```

Comment:
```suggestion
    ShapiroWilkTest(X::AbstractArray{<:Real}, SWc=SWCoeffs(length(X)))
```
I don't think `N1` should be exposed to the user if setting it to anything other than the default value results in an error when computing the p-value. Also, it's unusual for functions to have a keyword argument denoting whether something is sorted; calling `issorted(X)` unconditionally seems fine to me.

Reply:
I made `N1` available here so that future implementation doesn't change the signature; But now I know it won't since `ShapiroWilkTest(X, SWc, N1=length(X))` produces two methods ;)

`is_sorted` allows for micro optimization if the sample is already sorted:
the whole S-W test with precomputed coeffs and sorted sample takes less time than checking if the sample is sorted:

```julia
using Random
Random.seed!(1)
k = 5000
X = sort(randn(k));
using BenchmarkTools
@assert issorted(X)
@btime issorted($X)
# 6.481 μs (0 allocations: 0 bytes)
```
(vs. ~`5µs` for SW test)